monai/data/utils.py

```diff
@@ -945,6 +948,55 @@ def __init__(self, even_divisible: bool = True, *args, **kwargs):
             self.total_size = data_len
 
 
+class DistributedWeightedRandomSampler(DistributedSampler):
+    """
+    Extend the `DistributedSampler` to support weighted sampling.
+    Refer to `torch.utils.data.WeightedRandomSampler`, for more details please check:
+    https://github.com/pytorch/pytorch/blob/master/torch/utils/data/sampler.py#L150
+
+    Args:
+        weights: a sequence of weights, not necessary summing up to one, length should exactly
+            match the full dataset.
+        num_samples_per_rank: number of samples to draw for every rank, sample from
+            the distributed subset of dataset.
+            if None, default to the length of dataset split by DistributedSampler.
+        replacement: if ``True``, samples are drawn with replacement, otherwise, they are
+            drawn without replacement, which means that when a sample index is drawn for a row,
+            it cannot be drawn again for that row, default to True.
+        generator: PyTorch Generator used in sampling.
+        even_divisible: if False, different ranks can have different data length.
+            for example, input data: [1, 2, 3, 4, 5], rank 0: [1, 3, 5], rank 1: [2, 4].'
+        args: additional arguments for `DistributedSampler` super class.
+        kwargs: additional arguments for `DistributedSampler` super class.
+
+    """
+
+    def __init__(
+        self,
+        weights: Sequence[float],
+        num_samples_per_rank: Optional[int] = None,
+        replacement: bool = True,
+        generator: Optional[torch.Generator] = None,
+        even_divisible: bool = True,
+        *args,
```

Comment:
It's getting very lengthy but it would much better for the position and keyword arguments here and in the inherited sampler to be explicitly declared rather than using args/kwargs. From a user's perspective it's not clear what these arguments are supposed to be without reading deeper into our documentation and then having to dig up Pytorch documentation as well. This would also allow us to put our position arguments after those required from the Pytorch type which I feel makes more sense, ie. additional arguments come after those "inherited" from the supertype's constructor.

Reply:
Hi @ericspod ,

Thanks for your suggestion.
The reason why we used `args, kwargs` for MONAI `DistributedSampler` and `DataLoader` it that:
Now we support PyTorch 1.5 - 1.8, but the args of `DistributedSampler` and `DataLoader` are slightly different in different PyTorch versions.
Maybe we can try to inherit the args when we drop support for 1.5 and 1.6 later.

Thanks.